guidelines/configuration-api-design.md

```diff
@@ -0,0 +1,36 @@
+## Designing configuration API's
+
+All configuration API's should be code first since that allows us to best guide the users and allows us to evolve the API's using obsolete messages.
+
+Configuration API's should be variable based and not returning `this`.
+
+Example:
+
+```
+var myConfig = endpointConfig.SomethingNeedingConfig()
+
+myConfig.SomeOption(X);
+myConfig.SomeOtherOption(X);
+
+```
+
+We prefer this over lambda based API's since:
+
+1. Allows us to use lambdas for config options where we do make use of delayed execution.
+2. Users are more comforable with this type of API
+3. Nested configuration options becomes harder to read with lambdas
+4. It can confuse users as to when the lambda actually gets executed. Variable scoping, can I call a DB? etc
+5. Most of our current apis (Transport, Persistence, etc) is variable based
+
+### Use lambdas where execution is delayed
+
+If users needs to provide customizations that will be used later prefer to use a lambda.
+
+Example:
+
+```
+var transportConfig = endpointConfig.UseTransport<MsmqTransport>();
+
+transportConfig.MsmqLabelGenerator(context => return $"{context.Headers['NServiceBus.EnclosedMessageTypes']}");
+
+```
```

Comment:
should this really be a file specific for each design guideline instead of a single "design guidelines" file? (and the "performance" file could be "coding guidelines"?)